From 843ca079462dd394c10d3fced5df85f20cfd5dab Mon Sep 17 00:00:00 2001
From: Shivani Tambatkar <quic_stambatk@quicinc.com>
Date: Mon, 4 Dec 2023 12:13:40 -0800
Subject: [PATCH 1/7] hostapd: add support for device bandwidth parameters

Add new parameters to structures hostapd_config and hostapd_freq_params
to store device bandwidth. Also modify hostapd_set_freq_params() to
include these parameters.

Signed-off-by: Shivani Tambatkar <quic_stambatk@quicinc.com>
Signed-off-by: Aloka Dixit <quic_alokad@quicinc.com>
---
 src/ap/ap_config.h              |  2 ++
 src/ap/ap_drv_ops.c             | 12 ++++++++----
 src/ap/ap_drv_ops.h             |  6 ++++--
 src/ap/beacon.c                 |  4 +++-
 src/ap/dfs.c                    | 17 ++++++++++++-----
 src/ap/hostapd.c                | 12 ++++++++++--
 src/common/hw_features_common.c |  5 ++++-
 src/common/hw_features_common.h |  3 ++-
 src/drivers/driver.h            | 12 ++++++++++++
 wpa_supplicant/mesh.c           | 12 ++++++++----
 wpa_supplicant/wpa_supplicant.c |  2 +-
 11 files changed, 66 insertions(+), 21 deletions(-)

--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -1271,6 +1271,8 @@ struct hostapd_config {
 
 	int use_ru_puncture_dfs;
 	int ccfs;
+	int bandwidth_device;
+	int center_freq_device;
 };
 
 
--- a/src/ap/ap_drv_ops.c
+++ b/src/ap/ap_drv_ops.c
@@ -712,7 +712,8 @@ int hostapd_set_freq(struct hostapd_data
 		     int he_enabled, bool eht_enabled,
 		     int sec_channel_offset, int oper_chwidth,
 		     int center_segment0, int center_segment1,
-		     u16 ru_punct_bitmap, u8 ru_punct_ofdma)
+		     u16 ru_punct_bitmap, u8 ru_punct_ofdma,
+		     int bandwidth_device, int center_freq_device)
 {
 	struct hostapd_freq_params data;
 	struct hostapd_hw_modes *cmode = hapd->iface->current_mode;
@@ -730,7 +731,8 @@ int hostapd_set_freq(struct hostapd_data
 				    NULL, hostapd_get_punct_bitmap(hapd),
 				    hapd->iconf->he_6ghz_reg_pwr_type,
 				    ru_punct_bitmap,
-				    ru_punct_ofdma))
+				    ru_punct_ofdma,
+				    bandwidth_device, center_freq_device))
 		return -1;
 
 	if (hapd->driver == NULL)
@@ -1071,7 +1073,8 @@ int hostapd_start_dfs_cac(struct hostapd
 			  int sec_channel_offset, int oper_chwidth,
 			  int center_segment0, int center_segment1,
 			  bool radar_background,
-			  u16 ru_punct_bitmap, u8 ru_punct_ofdma)
+			  u16 ru_punct_bitmap, u8 ru_punct_ofdma,
+			  int bandwidth_device, int center_freq_device)
 {
 	struct hostapd_data *hapd = iface->bss[0];
 	struct hostapd_freq_params data;
@@ -1099,7 +1102,8 @@ int hostapd_start_dfs_cac(struct hostapd
 				    hostapd_get_punct_bitmap(hapd),
 				    hapd->iconf->he_6ghz_reg_pwr_type,
 				    ru_punct_bitmap,
-				    ru_punct_ofdma)) {
+				    ru_punct_ofdma,
+				    bandwidth_device, center_freq_device)) {
 		wpa_printf(MSG_ERROR, "Can't set freq params");
 		return -1;
 	}
--- a/src/ap/ap_drv_ops.h
+++ b/src/ap/ap_drv_ops.h
@@ -81,7 +81,8 @@ int hostapd_set_freq(struct hostapd_data
 		     int ht_enabled, int vht_enabled, int he_enabled,
 		     bool eht_enabled, int sec_channel_offset, int oper_chwidth,
 		     int center_segment0, int center_segment1,
-		     u16 ru_punct_bitmap, u8 ru_punct_ofdma);
+		     u16 ru_punct_bitmap, u8 ru_punct_ofdma,
+		     int bandwidth_device, int center_freq_device);
 int hostapd_set_rts(struct hostapd_data *hapd, int rts);
 int hostapd_set_frag(struct hostapd_data *hapd, int frag);
 int hostapd_sta_set_flags(struct hostapd_data *hapd, u8 *addr,
@@ -147,7 +148,8 @@ int hostapd_start_dfs_cac(struct hostapd
 			  int sec_channel_offset, int oper_chwidth,
 			  int center_segment0, int center_segment1,
 			  bool radar_background,
-			  u16 ru_punct_bitmap, u8 ru_punct_ofdma);
+			  u16 ru_punct_bitmap, u8 ru_punct_ofdma,
+			  int bandwidth_device, int center_freq_device);
 int hostapd_drv_do_acs(struct hostapd_data *hapd);
 int hostapd_drv_update_dh_ie(struct hostapd_data *hapd, const u8 *peer,
 			     u16 reason_code, const u8 *ie, size_t ielen);
--- a/src/ap/dfs.c
+++ b/src/ap/dfs.c
@@ -1049,7 +1049,8 @@ int hostapd_handle_dfs(struct hostapd_if
 		hostapd_get_oper_centr_freq_seg0_idx(iface->conf),
 		hostapd_get_oper_centr_freq_seg1_idx(iface->conf),
 		dfs_use_radar_background(iface),
-		iface->conf->ru_punct_bitmap, iface->conf->ru_punct_ofdma);
+		iface->conf->ru_punct_bitmap, iface->conf->ru_punct_ofdma,
+		iface->conf->bandwidth_device, iface->conf->center_freq_device);
 
 	if (res) {
 		wpa_printf(MSG_ERROR, "DFS start_dfs_cac() failed, %d", res);
@@ -1157,7 +1158,9 @@ static int hostapd_dfs_request_channel_s
 				      hostapd_get_punct_bitmap(iface->bss[0]),
 				      iface->conf->he_6ghz_reg_pwr_type,
 				      iface->conf->ru_punct_bitmap,
-				      iface->conf->ru_punct_ofdma);
+				      iface->conf->ru_punct_ofdma,
+				      iface->conf->bandwidth_device,
+				      iface->conf->center_freq_device);
 
 	if (err) {
 		wpa_printf(MSG_ERROR,
@@ -1235,7 +1238,7 @@ static void hostapd_dfs_update_backgroun
 				  oper_centr_freq_seg0_idx,
 				  oper_centr_freq_seg1_idx, true,
 				  iface->conf->ru_punct_bitmap,
-				  iface->conf->ru_punct_ofdma)) {
+				  iface->conf->ru_punct_ofdma, 0, 0)) {
 		wpa_printf(MSG_ERROR, "DFS failed to start CAC offchannel");
 		iface->radar_background.channel = -1;
 		return;
@@ -1319,7 +1322,9 @@ static int hostapd_dfs_testmode_set_beac
 				      hostapd_get_punct_bitmap(iface->bss[0]),
 				      iface->conf->he_6ghz_reg_pwr_type,
 				      iface->conf->ru_punct_bitmap,
-				      iface->conf->ru_punct_ofdma);
+				      iface->conf->ru_punct_ofdma,
+				      iface->conf->bandwidth_device,
+				      iface->conf->center_freq_device);
 
 	if (err) {
 		wpa_printf(MSG_ERROR, "DFS failed to calculate CSA freq params");
@@ -1810,7 +1815,9 @@ int hostapd_dfs_radar_detected(struct ho
 						     dfs_use_radar_background(iface),
 						     iface->conf->ru_punct_bitmap |
 						     iface->radar_bit_pattern,
-						     iface->conf->ru_punct_ofdma);
+						     iface->conf->ru_punct_ofdma,
+						     iface->conf->bandwidth_device,
+						     iface->conf->center_freq_device);
 		}
 
 		return hostapd_dfs_request_channel_switch(
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -352,6 +352,8 @@ int hostapd_reload_config(struct hostapd
 			hapd->iconf,
 			hostapd_get_oper_centr_freq_seg1_idx(oldconf));
 		hapd->conf = newconf->bss[j];
+		hapd->iconf->bandwidth_device = oldconf->bandwidth_device;
+		hapd->iconf->center_freq_device = oldconf->center_freq_device;
 		hostapd_reload_bss(hapd);
 	}
 
@@ -2621,7 +2623,9 @@ static int hostapd_setup_interface_compl
 				     hostapd_get_oper_centr_freq_seg1_idx(
 					      hapd->iconf),
 				     hapd->iconf->ru_punct_bitmap,
-				     hapd->iconf->ru_punct_ofdma)) {
+				     hapd->iconf->ru_punct_ofdma,
+				     hapd->iconf->bandwidth_device,
+				     hapd->iconf->center_freq_device)) {
 			wpa_printf(MSG_ERROR, "Could not set channel for "
 				   "kernel driver");
 			goto fail;
@@ -4389,7 +4393,9 @@ static int hostapd_change_config_freq(st
 				    hostapd_get_punct_bitmap(hapd),
 				    hapd->iconf->he_6ghz_reg_pwr_type,
 				    hapd->iconf->ru_punct_bitmap,
-				    hapd->iconf->ru_punct_ofdma))
+				    hapd->iconf->ru_punct_ofdma,
+				    conf->bandwidth_device,
+				    conf->center_freq_device))
 		return -1;
 
 	switch (params->bandwidth) {
@@ -4760,6 +4766,11 @@ hostapd_switch_channel_fallback(struct h
 	iface->conf->ieee80211ac = freq_params->vht_enabled;
 	iface->conf->ieee80211ax = freq_params->he_enabled;
 	iface->conf->ieee80211be = freq_params->eht_enabled;
+	iface->conf->bandwidth_device = freq_params->bandwidth_device;
+	iface->conf->bandwidth_device = freq_params->bandwidth_device;
+	iface->conf->center_freq_device = freq_params->center_freq_device;
+	iface->conf->bandwidth_device = freq_params->bandwidth_device;
+
 
 	hostapd_set_oper_centr_freq_seg0_idx(iface->conf, seg0_idx);
 	hostapd_set_oper_centr_freq_seg1_idx(iface->conf, seg1_idx);
--- a/src/common/hw_features_common.c
+++ b/src/common/hw_features_common.c
@@ -491,7 +491,8 @@ int hostapd_set_freq_params(struct hosta
 			    struct eht_capabilities *eht_cap,
 			    u16 punct_bitmap,
 			    u8 reg_6g_pwr_mode,
-			    u16 ru_punct_bitmap, u8 ru_punct_ofdma)
+			    u16 ru_punct_bitmap, u8 ru_punct_ofdma,
+			    int bandwidth_device, int center_freq_device)
 {
 	enum oper_chan_width oper_chwidth_legacy;
 	u8 seg0_legacy, seg1_legacy;
@@ -514,6 +515,8 @@ int hostapd_set_freq_params(struct hosta
 	data->center_freq2 = 0;
 	data->ru_punct_bitmap = punct_bitmap;
 	data->ru_punct_ofdma = punct_ofdma;
+	data->bandwidth_device = bandwidth_device;
+	data->center_freq_device = center_freq_device;
 	if (oper_chwidth == CONF_OPER_CHWIDTH_80MHZ)
 		data->bandwidth = 80;
 	else if (oper_chwidth == CONF_OPER_CHWIDTH_160MHZ ||
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -944,6 +944,18 @@ struct hostapd_freq_params {
 	 * link_id: If >=0 indicates the link of the AP MLD to configure
 	 */
 	int link_id;
+
+	/**
+	 * bandwidth_device - Device bandwidth in MHz, minimum 40 MHz.
+	 * The member 'bandwidth' corresponds to the operating bandwidth.
+	 */
+	int bandwidth_device;
+
+	/**
+	 * freq_device - Device center frequency in MHz, must coincide with
+	 * one edge of the operating bandwidth.
+	 */
+	int center_freq_device;
 };
 
 /**
@@ -6843,6 +6855,8 @@ union wpa_event_data {
 		u16 punct_bitmap;
 		 u16 ru_punct_bitmap;
                 u8 ru_punct_ofdma;
+		u32 ch_width_device;
+		u32 cf_device;
 	} ch_switch;
 
 	/**
--- a/wpa_supplicant/mesh.c
+++ b/wpa_supplicant/mesh.c
@@ -235,7 +235,9 @@ static int wpas_mesh_update_freq_params(
 		    he_capab, NULL, 0,
 		    ifmsh->conf->he_6ghz_reg_pwr_type,
 		    ifmsh->conf->ru_punct_bitmap,
-		    ifmsh->conf->ru_punct_ofdma)) {
+		    ifmsh->conf->ru_punct_ofdma,
+		    ifmsh->conf->bandwidth_device,
+		    ifmsh->conf->center_freq_device)) {
 		wpa_printf(MSG_ERROR, "Error updating mesh frequency params");
 		wpa_supplicant_mesh_deinit(wpa_s, true);
 		return -1;
@@ -519,9 +521,11 @@ static int wpa_supplicant_mesh_init(stru
 		wpa_s->mesh_params->handle_dfs = true;
 	}
 
-	if (ssid->eht)
-                conf->ru_punct_bitmap = freq->ru_punct_bitmap;
-
+	if (ssid->eht) {
+		conf->ru_punct_bitmap = freq->ru_punct_bitmap;
+		conf->bandwidth_device = freq->bandwidth_device;
+		conf->center_freq_device = freq->center_freq_device;
+	}
 	bss->iconf = conf;
 	ifmsh->conf = conf;
 
--- a/wpa_supplicant/wpa_supplicant.c
+++ b/wpa_supplicant/wpa_supplicant.c
@@ -3195,6 +3195,8 @@ skip_to_6ghz:
 	}
 
 skip_80mhz:
+        freq->bandwidth_device = 0;
+        freq->center_freq_device = 0;
 	if (hostapd_set_freq_params(&vht_freq, mode->mode, freq->freq,
 				    freq->channel, ssid->enable_edmg,
 				    ssid->edmg_channel, freq->ht_enabled,
@@ -3205,7 +3207,9 @@ skip_80mhz:
 				    &mode->he_capab[ieee80211_mode],
 				    &mode->eht_capab[ieee80211_mode], 0,
 				    freq->he_6ghz_reg_pwr_type,
-				    0, 0) != 0)
+				    0, 0,
+				    freq->bandwidth_device,
+				    freq->center_freq_device) != 0)
 		return false;
 
 	*freq = vht_freq;
--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -2889,7 +2889,9 @@ static int __ieee802_11_set_beacon(struc
 				    hostapd_get_punct_bitmap(hapd),
 				    iconf->he_6ghz_reg_pwr_type,
 				    iconf->ru_punct_bitmap,
-				    iconf->ru_punct_ofdma) == 0) {
+				    iconf->ru_punct_ofdma,
+				    iconf->bandwidth_device,
+				    iconf->center_freq_device) == 0) {
 		freq.link_id = -1;
 #ifdef CONFIG_IEEE80211BE
 		if (hapd->conf->mld_ap)
--- a/src/common/hw_features_common.h
+++ b/src/common/hw_features_common.h
@@ -50,7 +50,8 @@ int hostapd_set_freq_params(struct hosta
 			    struct eht_capabilities *eht_cap,
 			    u16 punct_bitmap,
 			    u8 reg_6g_pwr_mode,
-			    u16 ru_punct_bitmap, u8 ru_punct_ofdma);
+			    u16 ru_punct_bitmap, u8 ru_punct_ofdma,
+			    int bandwidth_device, int center_freq_device);
 void set_disable_ht40(struct ieee80211_ht_capabilities *htcaps,
 		      int disabled);
 int ieee80211ac_cap_check(u32 hw, u32 conf);
--- a/src/ap/ubus.c
+++ b/src/ap/ubus.c
@@ -772,7 +772,9 @@ hostapd_switch_chan(struct ubus_context
 				hostapd_get_punct_bitmap(hapd),
 				hapd->iconf->he_6ghz_reg_pwr_type,
 				iconf->ru_punct_bitmap,
-				iconf->ru_punct_ofdma);
+				iconf->ru_punct_ofdma,
+				iconf->bandwidth_device,
+				iconf->center_freq_device);
 
 	for (i = 0; i < hapd->iface->num_bss; i++) {
 		struct hostapd_data *bss = hapd->iface->bss[i];
